Scripts moved between documents

hiroshige-g opened the pull request:
This follows the changes in whatwg/html#2673, but also tests the issue at whatwg/html#2137 in favor of the current spec.

Originally at https://github.com/web-platform-tests/wpt/pull/5911.

domenic committed:
Test moving <script>s between documents

This follows the changes in https://github.com/whatwg/html/pull/2673, but also tests the issue at https://github.com/whatwg/html/issues/2137 in favor of the current spec.

domenic committed:
Also test module scripts

domenic committed:
Make window.didExecute undefined, not false, for no execution, to avoid Firefox vs. others issue

domenic committed:
No need for timeout per annevk's review

domenic committed:
Further fixes per review and my own discoveries

Turns out the timeout is needed at least a bit

hiroshige-g committed:
Add failing fetch cases

foolip commented:
@jgraham I discovered via https://github.com/foolip/wpt-stats/commit/f594c4dd3a0962ff26fb7a6c72fdc71d8b265ee9/checks?check_suite_id=260736134 that Taskcluster hasn't started for this PR. Can you report this to the team?

@hiroshige-g I see this is just a draft and it might not matter, but if you want to retrigger you can either close and reopen the PR, or just push new commits.

foolip commented:
Weird, still no Taskcluster run here despite closing and reopening. Ping @jgraham.

domenic commented:
It seems very likely that WIP pull requests do not get CI run, so I'm going to try pushing the "ready to review" button (and maybe closing/reopening) to see what happens.

foolip commented:
@domenic that Taskcluster probably wasn't because this was a draft, but because the branch was based on https://github.com/web-platform-tests/wpt/commit/0d5ca044187f3e1122661048e84a54e736296198 before the Taskcluster config was in the repo. Just rebasing the branch ought to have done the trick, in other words.

domenic commented:
@hiroshige-g this still says "do not review". Is that still accurate, or should we try to merge it (and merge https://github.com/whatwg/html/pull/2673)?

I guess there are still some bot problems, even after rebasing. Maybe we should just create a new pull request with the same contents.

jgraham commented:
My expectation is that rebasing this again will fix the TC issues. If it doesn't that seems like a bug.

domenic commented:
Build problem fixed by rebasing. @hiroshige-g this still says "do not review". Is that still accurate, or should we try to merge it (and merge whatwg/html#2673)?

hiroshige-g commented:
A couple of things to do:

- Fix the expectation for failed fetch + moved to another Window. Currently the test expects error event not fired, but according to whatwg/html#2673 the error event should be fired.
- I feel its better to separate tests for error/load events (relatively minor issue) and tests for whether the script itself is executed or not (the primary thing to test).

Drafting...

hiroshige-g commented:
Er, also I noticed the tests should be serialized, because the tests touch `window.didExecute`, but `async_test`s are not serialized.

hiroshige-g commented:
... and if we should serialize, then probably we should split tests into separate files to avoid timeouts.
Alternatively, we can test concurrently in `<iframe>`s.

hiroshige-g committed:
Isolate tests within iframes, and expect error events for failed fetch.

hiroshige-g commented:
Now ready for review, PTAL.

- I moved subtests into `<iframe>`s for isolation while running multiple subtests in parallel.
- I separated subtests about load/error events from subtests about whether scripts themselves are evaluated or not.

hiroshige-g commented:
Maybe it's better to add my commits back to https://github.com/web-platform-tests/wpt/pull/5911 to preserve previous contexts.

domenic commented on html/semantics/scripting-1/the-script-element/resources/slow-flag-setter.py:
```diff
@@ -0,0 +1,15 @@
+    body = "window.didExecute = true;"
+  else:
+    headers.append(("Transfer-encoding", "chunked"))
+    body = "Invalid\n\Chunk\n"
```
Is the `\Chunk` intended here? I assume it doesn't affect the results but it looks potentially like a typo.

hiroshige-g replied:
Should be just a typo. Fixed.

domenic commented on html/semantics/scripting-1/the-script-element/moving-between-documents.html:
```diff
@@ -0,0 +1,132 @@
+            iframe.contentWindow,
+            iframe.contentDocument.implementation.createHTMLDocument("")]);
+      } else {
+        reject("Invalid document type: " + documentType);
```
```suggestion
        reject(new Error("Invalid document type: " + documentType));
```

domenic commented on html/semantics/scripting-1/the-script-element/moving-between-documents.html:
```diff
@@ -0,0 +1,132 @@
+          }, 0);
+        }
+
+        setTimeout(() => {
```
Use t.step_timeout here?

hiroshige-g replied:
There are two concurrent tests (`t` and `tEvent`). I preferred very slightly the current code:

```
setTimeout(() => {
  t.step_func_done(...)();
  tEvent.step_func_done(...)();
}, 2000);
```

over:

```
t.step_timeout(() => { ... }, 2000);
tEvent.step_timeout(() => { ... }, 2000);
```

to avoid a slight race condition between two step_timeout()s, but I'm still not so sure.
WDYT?

domenic replied:
I agree a single timeout is better. Does the global `step_timeout` work? https://web-platform-tests.org/writing-tests/testharness-api.html says

> For single-page tests, step_timeout is also available as a global function.

but I'm not sure if that means it only exists in single-page tests, or if it's primarily used by single-page tests.

Either way it's worth adding a comment explaining why we're using setTimeout and/or global step_timeout.

hiroshige-g replied:
Global `step_timeout()` looks good. Done (and added a comment).

domenic reviewed (approved):
Mostly LGTM. However seeing the test results makes me think we should fire the error events in all cases, sigh. I will discuss more on the spec PR.

hiroshige-g committed:
reject with an Error instead of a raw string

Co-Authored-By: Domenic Denicola <d@domenic.me>

hiroshige-g committed:
Fix a typo

hiroshige-g committed:
Nit: use \r\n for chunked encoding

hiroshige-g committed:
Use step_timeout

hiroshige-g committed:
Test Window error events

hiroshige-g committed:
Fix slow-flag-setter to handle parse-error

hiroshige-g committed:
Add tests for moving inline scripts after #prepare-a-script.

hiroshige-g committed:
More comprehensive testing.
- Tests are split into separate HTMLs, to avoid per-server connection limit
  and timeouts.
- Added fixes/workarounds.

domenic commented:
Can you rebase this on master? Maybe that will fix the stalled build pipeline.

hiroshige-g committed:
Remove console.log

hiroshige-g committed:
Fix: actually move the script back for move-back tests

hiroshige-g committed:
Use 404 instead of transfer-encoding error for fetch error

domenic committed:
Fix lint

domenic commented:
@hiroshige-g I pushed a commit to fix the lint errors, but it seems like there are still some TODO comments.

domenic commented:
Ping @hiroshige-g; no rush, but just wanted to check to make sure you didn't forget about this.

hiroshige-g committed:
Resolve TODOs

hiroshige-g committed:
Add "Do not manually edit" warning

hiroshige-g commented:
Fixed TODOs, aligning with the latest https://github.com/whatwg/html/pull/2673.

stephenmcgruer commented:
Failed wpt-firefox-nightly-stability due to too many tests (timed out), and weirdly so did wpt-chrome-dev-results (but before it had even downloaded Chrome...).

Admin-merging.
